rfd/0234-in-band-mfa-ssh-sessions.md

```diff
@@ -0,0 +1,755 @@
+    MFAPromptResponseReference reference = 1;
+  }
+}
+
+// MFAPromptResponseReference instructs the SSH service to retrieve the MFA response from the MFA service.
+message MFAPromptResponseReference {
+  // challenge_name is the name of the MFA challenge created by the client.
+  string challenge_name = 1;
+}
+
+// SessionPayload contains identification information about an SSH session.
+message SessionPayload {
+  // session_id is the SSH session hash computed from SSH session state. For example, in Go this would be the value from
+  // crypto/ssh#ConnMetadata.SessionID().
+  bytes session_id = 1;
+}
+```
+
+#### MFA Service
+
+A new gRPC service `MFAService` will be created to encapsulate MFA-related RPCs, separating them from the legacy
+`AuthService`. In the future, all new MFA-related RPCs should be added to this service and existing MFA-related RPCs in
+the `AuthService` should be gradually migrated to `MFAService`.
+
+A new service was opted to be created instead of extending/adding new RPCs to the existing `AuthService` to maintain
+clear separation of concerns and to avoid continuing to bloat the `AuthService` with more responsibilities and
+complexity. Additionally, the RPCs defined in this new service are specifically focused on MFA challenges for user
+sessions, instead of further expanding the existing `CreateAuthenticateChallenge` RPC which is more general-purpose.
+
+The RPCs will mirror the existing headless authentication flow. Like `GetHeadlessAuthentication`, the
+`VerifyValidatedMFAChallenge` method will create a watcher with a timeout to wait for the [ValidatedMFAChallenge
+resource](#storing-validated-mfa-responses) to exist in the backend, rather than requiring the client to poll repeatedly
+until the resource exists in the local backend or for replication to complete to the leaf backend.
+
+```proto
+package teleport.mfa.v1;
+
+// MFAService defines the Multi-Factor Authentication (MFA) service. While this service is currently focused on
+// user sessions, new MFA related RPCs should be added here instead of the AuthService, to maintain a clear separation
+// of concerns instead of further bloating the AuthService.
+service MFAService {
+  // CreateChallenge creates an MFA challenge that is tied to a user session.
+  rpc CreateChallenge(CreateChallengeRequest) returns (CreateChallengeResponse);
+  // ValidateChallenge validates the MFA challenge response for a user session and stores the validated response in the
+  // backend.
+  rpc ValidateChallenge(ValidateChallengeRequest) returns (ValidateChallengeResponse);
+  // ReplicateValidatedMFAChallenge replicates a validated MFA challenge to a leaf cluster for retrieval during SSH session
+  // establishment. It is a NOOP when used in the root cluster. It is called by the reverse tunnel server when a
+  // validated challenge is created for a leaf cluster.
+  rpc ReplicateValidatedMFAChallenge(ReplicateValidatedMFAChallengeRequest) returns (ValidatedMFAChallenge);
```

Comment:
bikeshed: `ReplicateValidatedMFAChallenge` might be useful or required by the proxy for its own cluster if it's doing proxy recording mode or some other kind of "mfa impersonation", so I propose `CreateValidatedMFAChallenge`. Besides, in the case of a leaf cluster this is a method called by a proxy that doesn't really carry any proof that the original MFA challenge even exists, so IMO it's clearer if it's just a create operation that is explicitly creating a pre-validated challenge.